Security/Http/Firewall/TwoFactorListener.php

```diff
@@ -158,12 +159,21 @@ public function handle(GetResponseEvent $event)
             return;
         }
 
+        $preferredProvider = $request->get('preferProvider');
+        if ($preferredProvider) {
+            try {
+                $currentToken->preferTwoFactorProvider($preferredProvider);
+            } catch (UnknownTwoFactorProviderException $e) {
+                // Bad user input
+            }
+        }
+
         // Let routes pass, e.g. if a route needs to be callable during two-factor authentication
-        if ($this->twoFactorAccessDecider->isAccessible($request, $currentToken)) {
+        if ($this->twoFactorAccessDecider->isAccessible($request, $currentToken) && !$preferredProvider) {
             return;
         }
 
-        if (!$this->isAuthFormRequest($request)) {
+        if (!$this->isAuthFormRequest($request) || $preferredProvider) {
```

Comment:
Alternative implementation: Dispatch the REQUIRE event, when the form is shown, not when the redirect happens. That's how I'm doing it in my WIP 5.x implementation. Because dispatching the event only on redirect becomes a problem when the login directly redirects to the user to the 2fa form. Then the REQUIRE event would never be dispatched.

Reply:
I've pushed my WIP implementation, have a look here: https://github.com/scheb/two-factor-bundle/blob/954c07de2e8992680ef88cd1114239e38941ae47/Security/Http/Firewall/TwoFactorListener.php#L166-L171

Reply:
I was thinking about moving the event, but I was not sure about the usage.

I'll take a look, thanks!

Reply:
So I put back the prefered provider set on the form and move the event trigger.

It works, but I have to refresh the page after changing the provider. The change is set after the listener, this is why I moved the logic on the listener directly.

I'm still investigating the issue, but I'm hearing for good advices. :+1:

Reply:
When you change the provider, you'd need to dispatch the REQUIRE event once again.

I see your point with moving the logic into the listener, that would make things easier. Unfortunately we cannot do that without a breaking change, since that method is an extension point and removing it from the controller within the 4.x branch may break people's integration.

Reply:
The only thing I can try is to listen to an another kernel event triggered after the controller.

I'll look at this.

Reply:
I find a way using the `kernel.response` event on `TwoFactorProviderPreparationListener`. In a nutshell, I just save the token on the require event and moved the process logic on the response event.

Is this method ok to you?

Reply:
It would work, but it feels wrong. It implicitly requires the events being processed in a certain order (which they do of course, because we're aware of the inner workings of the system), but 
the implementation doesn't enforce it and that makes it look "hacky".

Reply:
It works, currently using the fork on my project.

I understand it looks "hacky", but it's technically correct, we are just using two events.

I don't currently have any solution except this one and moving the `preferedProvider` catch out of the controller. May you have a better idea without BC break?

I think this solution is still ok waiting 5.x